assists/baremetalconfig_xlnx.py

```diff
@@ -298,55 +299,67 @@ def get_mapped_nodes(sdt, node_list, options):
-               access_phandle_list.extend(rsrc_grp_phanlde)
-            except KeyError:
-               pass
-
-            mapped_phandle = [handle for handle in access_phandle_list for phandle in all_phandles if handle == phandle]
-            if len(access_phandle_list) != len(mapped_phandle):
-               wrong_mapped_handle = [handle for handle in access_phandle_list if handle not in mapped_phandle]
-               wrong_nodes = [node.name for node in sdt.tree['/'].subnodes() for handle in wrong_mapped_handle if handle == node.phandle]
-               if wrong_nodes:
-                  print("[WARNING]: invalid node %s reference mentioned in access property please delete the references" % ','.join(wrong_nodes))
-
-            domain_valid_nodes = [node for node in node_list for handle in mapped_phandle if handle == node.phandle]
-            valid_nodes.extend(domain_valid_nodes)
-            # Remove duplicate nodes
-            valid_nodes = list(dict.fromkeys(valid_nodes))
-        return valid_nodes
-    else:
-        return valid_nodes
+                if domain_node['cpus'].value:
+                    print('ERROR: os,type property is missing in the domain', domain_node.name)
+            except:
+                pass
+
+        if bm_domain:
+            cpu_node = get_cpunode(sdt.tree, domain_node)
+            if re.search(cpu_node[0].name, match_cpunodes[0].parent.name):
+               # Baremetal access property
+               # Check for shared resources
+               try:
+                   try:
+                       shared_phandles.extend(domain_node['access'].value)
+                   except:
+                       pass
+                   rsc_domain = domain_node['include'].value
+                   for rsc in rsc_domain:
+                       match = [node for node in domain_nodes if node.phandle == rsc]
+                       if match:
+                           try:
+                               shared_phandles.extend(match[0]['access'].value)
+                           except:
+                               pass
+               except:
+                   pass
+            else:
+               try:
+                   for phandle in domain_node['access'].value:
+                       invalid_phandles.append(phandle)
+               except:
+                   pass
+        else:
+            try:
+                for phandle in domain_node['access'].value:
+                    invalid_phandles.append(phandle)
+            except:
+                pass
+
+    # Remove duplicate phandle
+    all_phandles = list(dict.fromkeys(all_phandles))
+    # Remove shared phandles from invalid phandles list
+    invalid_phandles_list = [phandle for phandle in invalid_phandles if phandle not in shared_phandles]
+    valid_phandles = [phandle for phandle in all_phandles if phandle not in invalid_phandles_list]
+    valid_nodes = [node for node in node_list for handle in valid_phandles if handle == node.phandle]
```

Comment:
Rather than making this last block as compact as possible. We'd be better to just iterate and use pnode() calls via LopperTree.

Alternatively, this should all go into a lopper_lib routine that returns all referenced nodes from a domain, that implies that they are valid phandles and can be found via some sort of reference.

What are the valid_nodes being used for ?

Reply:
Valid nodes list is being generated based on the below assumptions as per system device-tree spec 
1) without any domains cpu cluster address-map property mapped nodes are valid nodes
2) with domains below are assumptions 
	i)  if the domain node has access property then it takes priority than address-map property (i.e if any other domain node contains access property and the same peripheral node is not mapped in our domain access property we should remove that node valid_node list) 
      ii) if the same peripheral node is mapped in our domain and other domain access property we should include it in the valid_node list 
     iii) if the peripheral node is in a shared resource and if it is mapped for the our domain include it in the valid node list else remove it from the valid node list.

Reply:
Aha. I see. This is something that eventually was planned for the lopper_lib() routines, and we'd be better off starting to split things out and into those routines. We need to have a "allowed_devices() or visible_devices()", that we can simply call with a domain and get the list of nodes and their memory location. Alternatively, we need a single overall domain routine that builds a dictionary for each domain, and then we can just access it via a specific domain name.

About point #1, are we really supporting a mode without domains ? Without domains, the type of modification we are talking about should just be a lop to select and modify/delete/rename.

And for 2.1, can you provide an example ? The description doesn't ready properly to me.

Basically, we should be documenting these cases in the code itself, so I'm asking questions to drive us towards comments above the processing that document things :D

Reply:
Agree @zeddii , For 2.1 please find yaml content below 
# Peripheral Split Domain Example
# Ex: This example is for ZynqMP ZCU102 default petaliux bsp design system device-tree.
resource_groups:
    shared_dev:
        access: 
          - dev: csudma_0 
            flags: default
    shared_mem:
        memory:
          - start: 0x60000000
            size: 512M
            flags: default

domains:
    linux: 
        id: 0x0
        cpus:
          - cluster: cpus_a53
            cpumask: 0x3
            mode:
              secure: false
              el: 0x1
        access: 
          - dev: i2c0 
            flags: default
          - dev: i2c1 
            flags: default
          - dev: can1 
            flags: default
        memory: 
          - start: 0x0
            size: 1G
            flags: default
        include:
          - shared_mem
          - shared_dev
    bm_a53:
        id: 0x1
        cpus:
          - cluster: cpus_a53
            cpumask: 0x2
            mode:
              secure: false
              el: 0x3
        access:  
          - dev: i2c0 
            flags: default
          - dev: i2c1 
            flags: default
          - dev: lpd_dma_chan7 
            flags: default
        include:
          - shared_dev
    bm_r5:
        id: 0x2
        cpus:
          - cluster: cpus_r5
            cpumask: 0x2
            mode:
              secure: false
              el: 0x3
        access:  
          - dev: lpd_dma_chan7 
            flags: default
        memory:  
          - start: 0x40000000
            size: 512M
            flags: default
        include:
          - shared_dev
    bm_mb:
        id: 0x1
        cpus:
          - cluster: cpus_microblaze_1 
            cpumask: 0x1
            mode:
              secure: false
              el: 0x3
        access:  
          - dev: i2c0 
            flags: default
          - dev: i2c1 
            flags: default
          - dev: can1 
            flags: default
        include:
          - shared_dev
 
If i generate baremetal driver list for the R5 processor using above yaml iicps driver shouldn't be pulled as it is being used by the Linux domain and not mapped in the baremetal R5 domain, Similarly if we genrerate linux device-tree for the above yaml file lpd_dma_chan7 node shouldn't be there in the linux domain and memory node also should get updated as per the yaml mapping. 

I have tested changes on top of master-next branch will send one more pull request please review and let me know the review comments if needed will post v3...

Reply:
> Aha. I see. This is something that eventually was planned for the lopper_lib() routines, and we'd be better off starting to split things out and into those routines. We need to have a "allowed_devices() or visible_devices()", that we can simply call with a domain and get the list of nodes and their memory location. Alternatively, we need a single overall domain routine that builds a dictionary for each domain, and then we can just access it via a specific domain name.
> 
> About point #1, are we really supporting a mode without domains ? Without domains, the type of modification we are talking about should just be a lop to select and modify/delete/rename.
Yes default is without domains only, whatever the functionality that exists today are without domains, 
My patches added was to add support for domains on top of existing functionality (i mean we should support both with domains use and with domains use case)... 
> 
> And for 2.1, can you provide an example ? The description doesn't ready properly to me.
> 
> Basically, we should be documenting these cases in the code itself, so I'm asking questions to drive us towards comments above the processing that document things :D